types/fingerprintjs2/index.d.ts

```diff
@@ -1,46 +1,87 @@
+     get(callback: (components: FingerPrint2Component[]) => void): void;
+
+     x64hash128(value: string, num: number): string;
+   }
 }
 
-interface Fingerprint2Options {
-    swfContainerId?: string;
-    swfPath?: string;
-    userDefinedFonts?: string[];
-    excludeUserAgent?: boolean;
-    excludeLanguage?: boolean;
-    excludeColorDepth?: boolean;
-    excludeScreenResolution?: boolean;
-    excludeTimezoneOffset?: boolean;
-    excludeSessionStorage?: boolean;
-    excludeIndexedDB?: boolean;
-    excludeAddBehavior?: boolean;
-    excludeOpenDatabase?: boolean;
-    excludeCpuClass?: boolean;
-    excludePlatform?: boolean;
-    excludeDoNotTrack?: boolean;
-    excludeCanvas?: boolean;
-    excludeWebGL?: boolean;
-    excludeAdBlock?: boolean;
-    excludeHasLiedLanguages?: boolean;
-    excludeHasLiedResolution?: boolean;
-    excludeHasLiedOs?: boolean;
-    excludeHasLiedBrowser?: boolean;
-    excludeJsFonts?: boolean;
-    excludeFlashFonts?: boolean;
-    excludePlugins?: boolean;
-    excludeIEPlugins?: boolean;
-    excludeTouchSupport?: boolean;
-    excludePixelRatio?: boolean;
-    excludeHardwareConcurrency?: boolean;
+interface FingerPrint2Component {
+   key: string;
+   value: string;
 }
+interface Fingerprint2Options {
+   preprocessor?: (key: string, value: string) => void;
+   audio?: {
+     timeout: number;
+     excludeIOS11: boolean;
+   };
+   fonts?: {
+     swfContainerId?: string;
+     swfPath?: string;
+     userDefinedFonts?: string[];
+     extendedJsFonts?: boolean;
+   };
+
+   screen?: {
+     // To ensure consistent fingerprints when users rotate their mobile devices
+     detectScreenOrientation: boolean;
+   };
+
+   plugins?: {
+     sortPluginsFor: RegExp[];
+     excludeIE: boolean;
+   };
+
+   extraComponents?: [
+     { key: string; getData: (done: any, options: any) => void }
+   ];
+
+   excludes?: {
+     userAgent?: boolean;
+     language?: boolean;
+     colorDepth?: boolean;
+     deviceMemory?: boolean;
+     pixelRatio?: boolean;
+     hardwareConcurrency?: boolean;
+     screenResolution?: boolean;
+     availableScreenResolution?: boolean;
+     timezoneOffset?: boolean;
+     timezone?: boolean;
+     sessionStorage?: boolean;
+     localStorage?: boolean;
+     indexedDb?: boolean;
+     addBehavior?: boolean;
+     openDatabase?: boolean;
+     cpuClass?: boolean;
+     platform?: boolean;
+     doNotTrack?: boolean;
+     plugins?: boolean;
+     canvas?: boolean;
+     webgl?: boolean;
+     webglVendorAndRenderer?: boolean;
+     adBlock?: boolean;
+     hasLiedLanguages?: boolean;
+     hasLiedResolution?: boolean;
+     hasLiedOs?: boolean;
+     hasLiedBrowser?: boolean;
+     touchSupport?: boolean;
+     fonts?: boolean;
+     fontsFlash?: boolean;
+     audio?: boolean;
+     enumerateDevices?: boolean;
+   };
+ }
 
-export = Fingerprint2;
-export as namespace Fingerprint2;
+declare var Fingerprint2: Fingerprint2Js.Fingerprint2Interface;
```

Comment:
This change would mean that this package can no longer be imported as a module. However, this package supports UMD and can be imported as either a CommonJS or AMD module in addition to being loaded as a script.

The above removed lines are intended to illustrate that:

```ts
export = Fingerprint2;
export as namespace Fingerprint2;
```

The `export = Fingerprint2` statement indicates the CommonJS/AMD behavior, while `export as namespace Fingerprint2` indicates the script behavior.

Reply:
@rbuckton can you elaborate more on what is meant by your comment ?

What was @jwobbeoya supposed to change in the PR ? 

I don't quite follow what the desired outcome was or what the problem was with the PR.

I had previously updated this but I'm not entirely sure what the different export syntaxes do: 'export as', and "export = ...", "export as namespace ..."

Can you give a concrete example of both how this intended to be used when it is imported and how it should export ?

I had previously imported only by importing the script directly: 
import "fingerprintjs2"

but should you be able to import this particular library with this style import: 
import { Fingerprint2 } from "fingerprintjs2"
?

Please clarify as I wasn't able to follow what you were intending to communicate in your comment but I would like to understand.

Reply:
what if you have multiple exports in the fingerprintjs2/index.d.ts ?

Reply:
`export as namespace` is TypeScript syntax to indicate a UMD module's global name when referenced as a script instead of a module. When used, it adds an entry to the type system that _also_ makes the exports of the module available as a global variable.

If you had a package `foo` that was available either as a global or as a module, you would write:

```ts
// foo/index.d.ts
export const bar = 1;
export const baz = 2;
export as namespace foo;
```

And the consumer could write _either_ of the following:

```ts
// inside of a script
foo.bar; // available via global `foo`

// inside of a module
import * as foo from "foo";
foo.bar; // available as an import
```

The UMD module can have multiple exports, just like any other module. In this case, the fingerprintjs2 module was using `export =` syntax, which is a special syntax TypeScript uses to emulate a CommonJS `module.exports = ...` assignment. In this case, the thing that is exported is a TypeScript namespace. TypeScript namespaces support multiple exports as well:

```ts
namespace foo {
  export const bar = 1;
  export const baz = 1;
}
export = foo;
export as namespace foo;
```